central/node/datastore/store/postgres/cve_store.go

```diff
@@ -0,0 +1,228 @@
+	defer metrics.SetPostgresOperationDurationTime(time.Now(), ops.GetMany, "NodeCVEs")
+
+	// Check cache first
+	cachedCVEs, missingIDs := s.cache.GetMany(cveIDs)
+	idToCVEMap := make(map[string]*storage.NodeCVE, len(cveIDs))
+
+	// Add cached CVEs to result
+	for id, cve := range cachedCVEs {
+		idToCVEMap[id] = cve
+	}
+
+	// If all CVEs were found in cache, return early
+	if len(missingIDs) == 0 {
+		return idToCVEMap, nil
+	}
+
+	// Fetch missing CVEs from database
+	rows, err := tx.Query(ctx, "SELECT serialized FROM "+nodeCVEsTable+" WHERE id = ANY($1::text[])", missingIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "querying CVEs from database")
+	}
+	defer rows.Close()
+
+	dbCVEs := make(map[string]*storage.NodeCVE)
+	for rows.Next() {
+		var data []byte
+		if err := rows.Scan(&data); err != nil {
+			return nil, errors.Wrap(err, "scanning CVE data")
+		}
+		msg := &storage.NodeCVE{}
+		if err := msg.UnmarshalVTUnsafe(data); err != nil {
+			return nil, errors.Wrap(err, "unmarshaling CVE data")
+		}
+		dbCVEs[msg.GetId()] = msg
+		idToCVEMap[msg.GetId()] = msg
+	}
+
+	// Cache the newly fetched CVEs
+	if len(dbCVEs) > 0 {
+		s.cache.SetMany(dbCVEs)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, errors.Wrap(err, "iterating over CVE rows")
+	}
+
+	return idToCVEMap, nil
+}
+
+func (s *nodeCVEStoreImpl) CopyFromNodeCves(ctx context.Context, tx *postgres.Tx, objs ...*storage.NodeCVE) error {
+	copyCols := []string{
+		"id",
+		"cvebaseinfo_cve",
+		"cvebaseinfo_publishedon",
+		"cvebaseinfo_createdat",
+		"operatingsystem",
+		"cvss",
+		"severity",
+		"impactscore",
+		"snoozed",
+		"snoozeexpiry",
+		"orphaned",
+		"orphanedtime",
+		"serialized",
+	}
+
+	// Process CVEs in batches using slices.Chunk
+	for batch := range slices.Chunk(objs, batchSize) {
+		// Prepare data for this batch
+		inputRows := make([][]interface{}, 0, len(batch))
+		deletes := make([]string, 0, len(batch))
+
+		for _, obj := range batch {
+			serialized, marshalErr := obj.MarshalVT()
+			if marshalErr != nil {
+				return errors.Wrapf(marshalErr, "marshaling CVE %s", obj.GetId())
+			}
+
+			inputRows = append(inputRows, []interface{}{
+				obj.GetId(),
+				obj.GetCveBaseInfo().GetCve(),
+				protocompat.NilOrTime(obj.GetCveBaseInfo().GetPublishedOn()),
+				protocompat.NilOrTime(obj.GetCveBaseInfo().GetCreatedAt()),
+				obj.GetOperatingSystem(),
+				obj.GetCvss(),
+				obj.GetSeverity(),
+				obj.GetImpactScore(),
+				obj.GetSnoozed(),
+				protocompat.NilOrTime(obj.GetSnoozeExpiry()),
+				obj.GetOrphaned(),
+				protocompat.NilOrTime(obj.GetOrphanedTime()),
+				serialized,
+			})
+
+			deletes = append(deletes, obj.GetId())
+		}
+
+		// Copy does not upsert so have to delete first.
+		_, err := tx.Exec(ctx, "DELETE FROM "+nodeCVEsTable+" WHERE id = ANY($1::text[])", deletes)
+		if err != nil {
+			return errors.Wrap(err, "deleting CVEs before copy")
+		}
+
+		// Invalidate cache for deleted CVEs
+		if len(deletes) > 0 {
```

Comment:
nitpick: If we are using transactions here and `CopyFrom` fails, we will still keep CVEs listed in `deletes` in DB. Would it be better to move deletion from cache after successful `CopyFrom`?

**EDIT:** After noticing `SetMany` - I'm not sure why we are deleting all nodeCVEs to set them again. I think only calling `SetMany` should be sufficient.